File: python/pybmap/catalog.py
"""Bose device catalog — known BMAP-capable devices.

Source: https://downloads.bose.com/lookup.xml
USB VID 0x05a7 is shared by all Bose devices.

This module is the authoritative device registry. Discovery and
connection code reference it for product ID → device type mapping.
"""

from collections import namedtuple

# All Bose USB devices share this vendor ID.
BOSE_USB_VID = 0x05A7

# BMAP Bluetooth service UUID (SDP record).
BMAP_UUID = "00000000-deca-fade-deca-deafdecacaff"

BoseDevice = namedtuple("BoseDevice", [
    "product_id",   # USB PID / Bluetooth Modalias product ID
    "codename",     # Internal Bose codename (from firmware URLs)
    "name",         # Marketing product name
    "category",     # "headphones", "earbuds", or "speaker"
    "config",       # Library config key ("qc_ultra2", "qc35") or None
])


# ── Device Catalog ──────────────────────────────────────────────────────────
# All known BMAP-capable Bose devices. Entries with config=None are
# recognized but not yet supported — they serve as a roadmap for
# future protocol implementations.

CATALOG = {
    # Headphones
    0x4017: BoseDevice(0x4017, "kleos",     "QuietComfort 35",                "headphones", "qc35"),
    0x4020: BoseDevice(0x4020, "baywolf",   "QuietComfort 35 II",             "headphones", "qc35"),
    0x4024: BoseDevice(0x4024, "goodyear",  "Noise Cancelling Headphones 700","headphones", None),
    0x4061: BoseDevice(0x4061, "vedder",    "QuietComfort 45",                "headphones", None),
    0x4082: BoseDevice(0x4082, "wolverine", "QuietComfort Ultra Headphones",  "headphones", "qc_ultra2"),

    # Earbuds
    0x4060: BoseDevice(0x4060, "olivia",    "QuietComfort Earbuds II",        "earbuds", None),
    0x4063: BoseDevice(0x4063, "edith",     "Ultra Open Earbuds",             "earbuds", None),
    0x4075: BoseDevice(0x4075, "prince",    "QuietComfort Ultra Earbuds",     "earbuds", None),

    # Speakers
    0x402D: BoseDevice(0x402D, "revel",     "Home Speaker 300",               "speaker", None),
    0x402F: BoseDevice(0x402F, "lando",     "Portable Home Speaker",          "speaker", None),
    0x4039: BoseDevice(0x4039, "duran",     "SoundLink Flex",                 "speaker", None),
    0x403A: BoseDevice(0x403A, "gwen",      "SoundLink Revolve+ II",          "speaker", None),
    0x4066: BoseDevice(0x4066, "lonestarr", "SoundLink Max",                  "speaker", None),
    0x4073: BoseDevice(0x4073, "scotty",    "SoundLink Flex 2nd Gen",         "speaker", None),
}


def lookup_device(product_id):
    """Look up a Bose device by product ID.

    Args:
        product_id: USB PID or Bluetooth Modalias product ID (int).

    Returns:
        BoseDevice namedtuple, or None if unknown.
    """
    return CATALOG.get(product_id)


def known_devices():
    """All known BMAP-capable Bose devices.

    Returns:
        List of BoseDevice namedtuples.
    """
    return list(CATALOG.values())


def supported_devices():
    """Devices with active library support (config is not None).

    Returns:
        List of BoseDevice namedtuples.
    """
    return [d for d in CATALOG.values() if d.config is not None]


def is_supported(product_id):
    """Check if a product ID has an active library implementation.

    Args:
        product_id: USB PID or Bluetooth Modalias product ID (int).

    Returns:
        True if the device has a config implementation.
    """
    entry = CATALOG.get(product_id)
    return entry is not None and entry.config is not None


def usb_ids(product_id):
    """Get USB vendor/product ID pair for a known device.

    Args:
        product_id: Bose product ID (int).

    Returns:
        (vendor_id, product_id) tuple, or None if unknown.
    """
    if product_id in CATALOG:
        return (BOSE_USB_VID, product_id)
    return None


def modalias(product_id):
    """Generate a Bluetooth Modalias string for a known device.

    Args:
        product_id: Bose product ID (int).

    Returns:
        Modalias string like "bluetooth:v009Bp4082d0000", or None.
    """
    if product_id in CATALOG:
        return "bluetooth:v%04Xp%04Xd0000" % (BOSE_USB_VID, product_id)
    return None

File: rust/src/catalog.rs
//! Bose device catalog — known BMAP-capable devices.
//!
//! Source: <https://downloads.bose.com/lookup.xml>

/// All Bose USB devices share this vendor ID.
pub const BOSE_USB_VID: u16 = 0x05A7;

/// BMAP Bluetooth service UUID (SDP record).
pub const BMAP_UUID: &str = "00000000-deca-fade-deca-deafdecacaff";

/// Device category.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Category {
    Headphones,
    Earbuds,
    Speaker,
}

/// A known Bose BMAP device.
#[derive(Debug, Clone)]
pub struct BoseDevice {
    pub product_id: u16,
    pub codename: &'static str,
    pub name: &'static str,
    pub category: Category,
    /// Library config key, or None if not yet supported.
    pub config: Option<&'static str>,
}

/// All known BMAP-capable Bose devices.
pub const CATALOG: &[BoseDevice] = &[
    // Headphones
    BoseDevice { product_id: 0x4017, codename: "kleos",     name: "QuietComfort 35",                 category: Category::Headphones, config: Some("qc35") },
    BoseDevice { product_id: 0x4020, codename: "baywolf",   name: "QuietComfort 35 II",              category: Category::Headphones, config: Some("qc35") },
    BoseDevice { product_id: 0x4024, codename: "goodyear",  name: "Noise Cancelling Headphones 700", category: Category::Headphones, config: None },
    BoseDevice { product_id: 0x4061, codename: "vedder",    name: "QuietComfort 45",                 category: Category::Headphones, config: None },
    BoseDevice { product_id: 0x4082, codename: "wolverine", name: "QuietComfort Ultra Headphones",   category: Category::Headphones, config: Some("qc_ultra2") },
    // Earbuds
    BoseDevice { product_id: 0x4060, codename: "olivia",    name: "QuietComfort Earbuds II",         category: Category::Earbuds, config: None },
    BoseDevice { product_id: 0x4063, codename: "edith",     name: "Ultra Open Earbuds",              category: Category::Earbuds, config: None },
    BoseDevice { product_id: 0x4075, codename: "prince",    name: "QuietComfort Ultra Earbuds",      category: Category::Earbuds, config: None },
    // Speakers
    BoseDevice { product_id: 0x402D, codename: "revel",     name: "Home Speaker 300",                category: Category::Speaker, config: None },
    BoseDevice { product_id: 0x402F, codename: "lando",     name: "Portable Home Speaker",           category: Category::Speaker, config: None },
    BoseDevice { product_id: 0x4039, codename: "duran",     name: "SoundLink Flex",                  category: Category::Speaker, config: None },
    BoseDevice { product_id: 0x403A, codename: "gwen",      name: "SoundLink Revolve+ II",           category: Category::Speaker, config: None },
    BoseDevice { product_id: 0x4066, codename: "lonestarr", name: "SoundLink Max",                   category: Category::Speaker, config: None },
    BoseDevice { product_id: 0x4073, codename: "scotty",    name: "SoundLink Flex 2nd Gen",          category: Category::Speaker, config: None },
];

/// Look up a Bose device by product ID.
pub fn lookup_device(product_id: u16) -> Option<&'static BoseDevice> {
    CATALOG.iter().find(|d| d.product_id == product_id)
}

/// All devices with active library support.
pub fn supported_devices() -> Vec<&'static BoseDevice> {
    CATALOG.iter().filter(|d| d.config.is_some()).collect()
}

/// Check if a product ID has an active library implementation.
pub fn is_supported(product_id: u16) -> bool {
    lookup_device(product_id).map_or(false, |d| d.config.is_some())
}

/// Get USB vendor/product ID pair for a known device.
pub fn usb_ids(product_id: u16) -> Option<(u16, u16)> {
    if lookup_device(product_id).is_some() {
        Some((BOSE_USB_VID, product_id))
    } else {
        None
    }
}

/// Generate a Bluetooth Modalias string for a known device.
pub fn modalias(product_id: u16) -> Option<String> {
    if lookup_device(product_id).is_some() {
        Some(format!("bluetooth:v{:04X}p{:04X}d0000", BOSE_USB_VID, product_id))
    } else {
        None
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_lookup_known() {
        let dev = lookup_device(0x4082).unwrap();
        assert_eq!(dev.codename, "wolverine");
        assert_eq!(dev.config, Some("qc_ultra2"));
    }

    #[test]
    fn test_lookup_unknown() {
        assert!(lookup_device(0xFFFF).is_none());
    }

    #[test]
    fn test_is_supported() {
        assert!(is_supported(0x4082));
        assert!(is_supported(0x4020));
        assert!(!is_supported(0x4024)); // NCH 700, no config
        assert!(!is_supported(0xFFFF));
    }

    #[test]
    fn test_supported_devices() {
        let devs = supported_devices();
        assert!(devs.len() >= 2); // at least QC35 + QC Ultra 2
        assert!(devs.iter().all(|d| d.config.is_some()));
    }

    #[test]
    fn test_usb_ids() {
        let (vid, pid) = usb_ids(0x4082).unwrap();
        assert_eq!(vid, 0x05A7);
        assert_eq!(pid, 0x4082);
        assert!(usb_ids(0xFFFF).is_none());
    }

    #[test]
    fn test_modalias() {
        let m = modalias(0x4082).unwrap();
        assert_eq!(m, "bluetooth:v05A7p4082d0000");
        assert!(modalias(0xFFFF).is_none());
    }
}
